[Intelligence Effects] Refactor WKIntelligenceTextEffectCoordinator to more easily support multiple strategies of applying effects

rr-codes opened the pull request:
#### 76b5355c58db6d0c468c6f334594f479b8a884d3
<pre>
[Intelligence Effects] Refactor WKIntelligenceTextEffectCoordinator to more easily support multiple strategies of applying effects
<a href="https://bugs.webkit.org/show_bug.cgi?id=285967">https://bugs.webkit.org/show_bug.cgi?id=285967</a>
<a href="https://rdar.apple.com/142937853">rdar://142937853</a>

Reviewed by Abrar Rahman Protyasha.

Rewrite and proofreading perform replacements by repeatedly applying contiguous replacements to the text. However, Smart Reply replaces the entire text per each replacement.

To support this differing logic, the effect coordinator has its view logic extracted into a new type intended to be reused and composed into each type that conforms to the new
Coordinating protocol, instead of having a lot of conditional statements. A subsequent PR will create a new specialization of the protocol for smart replies.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView didBeginWritingToolsSession:contexts:]):
(-[WKWebView proofreadingSession:didReceiveSuggestions:processedRange:inContext:finished:]):
(-[WKWebView viewForIntelligenceTextEffectCoordinator:]):
(-[WKWebView intelligenceTextEffectCoordinator:rectsForProofreadingSuggestionsInRange:completion:]):
(-[WKWebView intelligenceTextEffectCoordinator:updateTextVisibilityForRange:visible:identifier:completion:]):
(-[WKWebView intelligenceTextEffectCoordinator:textPreviewsForRange:completion:]):
(-[WKWebView intelligenceTextEffectCoordinator:contentPreviewForRange:completion:]):
(-[WKWebView intelligenceTextEffectCoordinator:decorateReplacementsForRange:completion:]):
(-[WKWebView intelligenceTextEffectCoordinator:setSelectionForRange:completion:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.h:
* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.mm:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebKitSwift/WebKitSwift.h:
* Source/WebKit/WebKitSwift/WritingTools/IntelligenceTextEffectChunk.swift: Added.
(range):
(IntelligenceTextEffectChunk.hash(into:)):
* Source/WebKit/WebKitSwift/WritingTools/IntelligenceTextEffectViewManager.swift: Added.
(IntelligenceTextEffectViewManager.source):
(IntelligenceTextEffectViewManager.effectView):
(IntelligenceTextEffectViewManager.activePonderingEffect):
(IntelligenceTextEffectViewManager.activeReplacementEffect):
(IntelligenceTextEffectViewManager.hasActiveEffects):
(IntelligenceTextEffectViewManager.setupViewIfNeeded):
(IntelligenceTextEffectViewManager.destroyViewIfNeeded):
(IntelligenceTextEffectViewManager.setActivePonderingEffect(_:)):
(IntelligenceTextEffectViewManager.setActiveReplacementEffect(_:)):
(IntelligenceTextEffectViewManager.hideEffects):
(IntelligenceTextEffectViewManager.showEffects):
(IntelligenceTextEffectViewManager.removeActiveEffects):
(IntelligenceTextEffectViewManager.updateActivePonderingEffect(_:)):
(IntelligenceTextEffectViewManager.reset):
* Source/WebKit/WebKitSwift/WritingTools/WKIntelligenceReplacementTextEffectCoordinator.h: Copied from Source/WebKit/WebKitSwift/WebKitSwift.h.
* Source/WebKit/WebKitSwift/WritingTools/WKIntelligenceReplacementTextEffectCoordinator.swift: Renamed from Source/WebKit/WebKitSwift/WritingTools/WKIntelligenceTextEffectCoordinator.swift.
(contextRange):
(replacementQueue):
(onFlushCompletion):
(hasActiveEffects):
(characterDelta(forReceivedSuggestions:)):
(startAnimation(for:)):
(requestReplacement(withProcessedRange:finished:characterDelta:operation:)):
(flushReplacements):
(restoreSelection(_:)):
(hideEffects):
(showEffects):
(removeActiveEffects):
(startReplacementAnimation(using:)):
(reset):
(WKIntelligenceReplacementTextEffectCoordinator.textPreview(for:)):
(WKIntelligenceReplacementTextEffectCoordinator.updateTextChunkVisibility(_:visible:force:)):
(WKIntelligenceReplacementTextEffectCoordinator.updateTextChunkVisibility(_:visible:)):
(WKIntelligenceReplacementTextEffectCoordinator.performReplacementAndGeneratePreview(for:effect:remainder:)):
(WKIntelligenceReplacementTextEffectCoordinator.replacementEffectWillBegin(_:)):
(WKIntelligenceReplacementTextEffectCoordinator.flushRemainingReplacementsIfNeeded):
(WKIntelligenceReplacementTextEffectCoordinator.replacementEffectDidComplete(_:)):
(async(_:)):
(platformTextPreview(from:suggestionRects:)):
* Source/WebKit/WebKitSwift/WritingTools/WKIntelligenceTextEffectCoordinator.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
(TEST(WritingTools, IntelligenceTextEffectCoordinatorDelegate_RectsForProofreadingSuggestionsInRange)):
(TEST(WritingTools, IntelligenceTextEffectCoordinatorDelegate_UpdateTextVisibilityForRange)):
(TEST(WritingTools, IntelligenceTextEffectCoordinatorDelegate_TextPreviewsForRange)):

Canonical link: <a href="https://commits.webkit.org/288954@main">https://commits.webkit.org/288954@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/15a97f9c9091833d26e3a60f6e06f06519281b8e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84869 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/4594 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39259 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/90013 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/35922 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/86954 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/4684 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/12571 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/66048 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/35922 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87914 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/3574 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/77126 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/46318 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/3455 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/31351 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/34995 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/74277 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32158 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/91386 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12208 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/8920 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/74534 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/12438 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72936 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/73659 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18034 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/16483 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/3882 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13224 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12160 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/17606 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/11995 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/15489 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13740 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/WebKitSwift/WritingTools/IntelligenceTextEffectChunk.swift:
```diff
@@ -0,0 +1,50 @@
+//
+// Copyright (C) 2025 Apple Inc. All rights reserved.
+//
```
This is not our complete copyright header.

pxlcoder reviewed (commented):
> contigious replacements to the text. However, smart replies replaces 

contigious -> contiguous
smart replies replaces -> "Smart Reply replaces" or "smart replies replace"

aprotyas commented on Source/WebKit/WebKitSwift/WritingTools/WKIntelligenceReplacementTextEffectCoordinator.h:
```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
```
2025

hortont424 commented on Source/WebKit/WebKitSwift/WritingTools/IntelligenceTextEffectViewManager.swift:
```diff
@@ -0,0 +1,195 @@
+
+    private var effectView: PlatformIntelligenceTextEffectView<Source>? = nil
+
+    // Use the corresponding setter functions instead of setting these directly.
```
not something you can enforce instead of leaving this comment? they are private, isn't that enough?

rr-codes replied:
(everything in this file is basically just copy-pasted from the original file) but yes now that it's own type and they are private I can probably just remove the comment

rr-codes replied:
(everything in this file is basically just copy-pasted from the original file) but yes now that it's own type and they are private I can probably just remove the comment

hortont424 commented on Source/WebKit/WebKitSwift/WritingTools/IntelligenceTextEffectViewManager.swift:
```diff
@@ -0,0 +1,195 @@
+
+    func setActiveReplacementEffect(_ effect: PlatformIntelligenceReplacementTextEffect<IntelligenceTextEffectChunk>?) async {
+        guard (self.activeReplacementEffect == nil && effect != nil) || (self.activeReplacementEffect != nil && effect == nil) else {
+            assertionFailure("Intelligence text effect coordinator: trying to either set a new replacement effect when there is an ongoing one, or trying to remove an effect when there are none.")
```
what variety of assertion is this? debug? release?

rr-codes replied:
assert and assertionFailure are debug asserts. precondition, preconditionFailure, and fatalError are release asserts.

rr-codes replied:
https://developer.apple.com/documentation/swift/debugging-and-reflection (the specifics are here)

hortont424 commented on Source/WebKit/WebKitSwift/WritingTools/IntelligenceTextEffectViewManager.swift:
```diff
@@ -0,0 +1,195 @@
+            return
+        }
+
+        // On macOS, the effect view is incapable of scrolling with the web view. Therefore, the effects need to be suppressed
```
This is not categorically true (we have a scrolling tree and can move things around in the UI process), we just haven't put in the effort. "Currently doesn't scroll with the web content" (the web view doesn't scroll) might be more accurate.

rr-codes committed:
[Intelligence Effects] Refactor WKIntelligenceTextEffectCoordinator to more easily support multiple strategies of applying effects
https://bugs.webkit.org/show_bug.cgi?id=285967
rdar://142937853

Reviewed by Abrar Rahman Protyasha.

Rewrite and proofreading perform replacements by repeatedly applying contiguous replacements to the text. However, Smart Reply replaces the entire text per each replacement.

To support this differing logic, the effect coordinator has its view logic extracted into a new type intended to be reused and composed into each type that conforms to the new
Coordinating protocol, instead of having a lot of conditional statements. A subsequent PR will create a new specialization of the protocol for smart replies.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView didBeginWritingToolsSession:contexts:]):
(-[WKWebView proofreadingSession:didReceiveSuggestions:processedRange:inContext:finished:]):
(-[WKWebView viewForIntelligenceTextEffectCoordinator:]):
(-[WKWebView intelligenceTextEffectCoordinator:rectsForProofreadingSuggestionsInRange:completion:]):
(-[WKWebView intelligenceTextEffectCoordinator:updateTextVisibilityForRange:visible:identifier:completion:]):
(-[WKWebView intelligenceTextEffectCoordinator:textPreviewsForRange:completion:]):
(-[WKWebView intelligenceTextEffectCoordinator:contentPreviewForRange:completion:]):
(-[WKWebView intelligenceTextEffectCoordinator:decorateReplacementsForRange:completion:]):
(-[WKWebView intelligenceTextEffectCoordinator:setSelectionForRange:completion:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.h:
* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.mm:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebKitSwift/WebKitSwift.h:
* Source/WebKit/WebKitSwift/WritingTools/IntelligenceTextEffectChunk.swift: Added.
(range):
(IntelligenceTextEffectChunk.hash(into:)):
* Source/WebKit/WebKitSwift/WritingTools/IntelligenceTextEffectViewManager.swift: Added.
(IntelligenceTextEffectViewManager.source):
(IntelligenceTextEffectViewManager.effectView):
(IntelligenceTextEffectViewManager.activePonderingEffect):
(IntelligenceTextEffectViewManager.activeReplacementEffect):
(IntelligenceTextEffectViewManager.hasActiveEffects):
(IntelligenceTextEffectViewManager.setupViewIfNeeded):
(IntelligenceTextEffectViewManager.destroyViewIfNeeded):
(IntelligenceTextEffectViewManager.setActivePonderingEffect(_:)):
(IntelligenceTextEffectViewManager.setActiveReplacementEffect(_:)):
(IntelligenceTextEffectViewManager.hideEffects):
(IntelligenceTextEffectViewManager.showEffects):
(IntelligenceTextEffectViewManager.removeActiveEffects):
(IntelligenceTextEffectViewManager.updateActivePonderingEffect(_:)):
(IntelligenceTextEffectViewManager.reset):
* Source/WebKit/WebKitSwift/WritingTools/WKIntelligenceReplacementTextEffectCoordinator.h: Copied from Source/WebKit/WebKitSwift/WebKitSwift.h.
* Source/WebKit/WebKitSwift/WritingTools/WKIntelligenceReplacementTextEffectCoordinator.swift: Renamed from Source/WebKit/WebKitSwift/WritingTools/WKIntelligenceTextEffectCoordinator.swift.
(contextRange):
(replacementQueue):
(onFlushCompletion):
(hasActiveEffects):
(characterDelta(forReceivedSuggestions:)):
(startAnimation(for:)):
(requestReplacement(withProcessedRange:finished:characterDelta:operation:)):
(flushReplacements):
(restoreSelection(_:)):
(hideEffects):
(showEffects):
(removeActiveEffects):
(startReplacementAnimation(using:)):
(reset):
(WKIntelligenceReplacementTextEffectCoordinator.textPreview(for:)):
(WKIntelligenceReplacementTextEffectCoordinator.updateTextChunkVisibility(_:visible:force:)):
(WKIntelligenceReplacementTextEffectCoordinator.updateTextChunkVisibility(_:visible:)):
(WKIntelligenceReplacementTextEffectCoordinator.performReplacementAndGeneratePreview(for:effect:remainder:)):
(WKIntelligenceReplacementTextEffectCoordinator.replacementEffectWillBegin(_:)):
(WKIntelligenceReplacementTextEffectCoordinator.flushRemainingReplacementsIfNeeded):
(WKIntelligenceReplacementTextEffectCoordinator.replacementEffectDidComplete(_:)):
(async(_:)):
(platformTextPreview(from:suggestionRects:)):
* Source/WebKit/WebKitSwift/WritingTools/WKIntelligenceTextEffectCoordinator.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
(TEST(WritingTools, IntelligenceTextEffectCoordinatorDelegate_RectsForProofreadingSuggestionsInRange)):
(TEST(WritingTools, IntelligenceTextEffectCoordinatorDelegate_UpdateTextVisibilityForRange)):
(TEST(WritingTools, IntelligenceTextEffectCoordinatorDelegate_TextPreviewsForRange)):

Canonical link: https://commits.webkit.org/288954@main
